Filedownloads url - Adhere to w3c standards

krmahadevan opened the pull request:
Fixes: #11466, #11458


New format

`POST /se/files`


Request:

`{ "filename": "file" }`


Response:
```
{
	"value": {
	"filename": "filename",
	"contents": "ContentsGoHere"
	}
}
```

`GET /se/files`

Request:
NONE

Response:
```
{
	"value": {
		"files":[
		"1",
		"2",
		"3"
		]
	}
}
```

**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

krmahadevan commented:
@diemol @titusfortner 

This PR takes care of ONLY streamlining the url to be compliant with `w3c` urls and also adds the capability to list files for download.

The PR for https://github.com/SeleniumHQ/selenium/issues/11443 will ensure that the downloads path references are all removed.

I thought it would be good if I go at this requirement on a piece by piece

surajverma255 commented:
Hi Community,
Have just started to explore the Open Source Community.
As a Noobie, I feel we should have a quick description of the changes someone is pushing which will make it easier to understand the context of change :)

titusfortner commented:
@krmahadevan I agree with going piece by piece, so I broke up the issue above into 5 components, this being two I think. Will review soon.

@surajverma255 welcome! Please read the issues linked in the PR to better understand the context.

titusfortner commented:
yeah, ok, @diemol will have to review the code as I'm lost in this part of Selenium, but the syntax looks correct to me.

Diego previously proposed to add file timestamp info with the GET response. To be honest, it would be a lot easier not to include it. Especially if we are already breaking things out by session, but we should agree on that bit.

diemol commented on java/src/org/openqa/selenium/grid/node/local/LocalNode.java:
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+      List<String> collected = Arrays.stream(
+        Optional.ofNullable(dir.listFiles()).orElse(new File[]{})
+      ).map(File::getName).collect(Collectors.toList());
+      ImmutableMap<String, Object> data = ImmutableMap.of("filenames", collected);
```
```suggestion
      ImmutableMap<String, Object> data = ImmutableMap.of("names", collected);
```

diemol commented on java/src/org/openqa/selenium/grid/node/local/LocalNode.java:
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+    String raw = string(req);
+    if (raw.isEmpty()) {
+      throw new WebDriverException(
+        "Please specify file to download in payload as {\"filename\": \"fileToDownload\"}");
```
```suggestion
        "Please specify file to download in payload as {\"name\": \"fileToDownload\"}");
```

diemol commented on java/src/org/openqa/selenium/grid/node/local/LocalNode.java:
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+    }
+
+    Map<String, Object> incoming = JSON.toType(raw, Json.MAP_TYPE);
+    String filename = Optional.ofNullable(incoming.get("filename"))
```
```suggestion
    String filename = Optional.ofNullable(incoming.get("name"))
```

diemol commented on java/src/org/openqa/selenium/grid/node/local/LocalNode.java:
```diff
@@ -495,7 +497,28 @@ public HttpResponse downloadFile(HttpRequest req, SessionId id) {
+      .map(Object::toString)
+      .orElseThrow(
+        () -> new WebDriverException(
+          "Please specify file to download in payload as {\"filename\": \"fileToDownload\"}"));
```
```suggestion
          "Please specify file to download in payload as {\"name\": \"fileToDownload\"}"));
```

diemol reviewed (approved):
Just very minor comments, as in the endpoint already shows it is about files. 

Thank you, @krmahadevan!

krmahadevan commented:
> Just very minor comments, as in the endpoint already shows it is about files.
> 
> Thank you, @krmahadevan!

Thank you @diemol. i have fixed the review comments as an additional commit and also rebased off of `trunk`. Can I go ahead and merge this or would you or @titusfortner do that? First PR after a committer :) So I dont know yet if I can hit the `merge` button (Even though I can see it.. :D )

diemol commented:
@krmahadevan sure, go for it.
